configure.ac

```diff
@@ -5729,127 +5732,157 @@ then
-    AC_DEFINE(HAVE_RL_COMPLETION_APPEND_CHARACTER, 1,
-      [Define if you have readline 2.2]),,
-    [
-#include <stdio.h> /* Must be first for Gnu Readline */
-#ifdef WITH_EDITLINE
-# include <editline/readline.h>
-#else
-# include <readline/readline.h>
-#endif
-    ])
-  AC_CHECK_DECL(rl_completion_suppress_append,
-    AC_DEFINE(HAVE_RL_COMPLETION_SUPPRESS_APPEND, 1,
-      [Define if you have rl_completion_suppress_append]),,
-    [
-#include <stdio.h> /* Must be first for Gnu Readline */
-#ifdef WITH_EDITLINE
-# include <editline/readline.h>
-#else
-# include <readline/readline.h>
-#endif
-    ])
+AS_VAR_IF([py_cv_lib_readline], [yes], [
+  WITH_SAVE_ENV([
+    CPPFLAGS="$READLINE_CFLAGS $CFLAGS"
+    LIBS="$READLINE_LIBS $LIBS"
 
-  # check for readline 4.0
-  AC_CHECK_LIB($LIBREADLINE, rl_pre_input_hook,
-    AC_DEFINE(HAVE_RL_PRE_INPUT_HOOK, 1,
-      [Define if you have readline 4.0]),,$READLINE_LIBS)
-
-  # also in 4.0
-  AC_CHECK_LIB($LIBREADLINE, rl_completion_display_matches_hook,
-    AC_DEFINE(HAVE_RL_COMPLETION_DISPLAY_MATCHES_HOOK, 1,
-      [Define if you have readline 4.0]),,$READLINE_LIBS)
-
-  # also in 4.0, but not in editline
-  AC_CHECK_LIB($LIBREADLINE, rl_resize_terminal,
-    AC_DEFINE(HAVE_RL_RESIZE_TERMINAL, 1,
-      [Define if you have readline 4.0]),,$READLINE_LIBS)
-
-  # check for readline 4.2
-  AC_CHECK_LIB($LIBREADLINE, rl_completion_matches,
-    AC_DEFINE(HAVE_RL_COMPLETION_MATCHES, 1,
-      [Define if you have readline 4.2]),,$READLINE_LIBS)
-
-  # also in readline 4.2
-  AC_CHECK_DECL(rl_catch_signals,
-    AC_DEFINE(HAVE_RL_CATCH_SIGNAL, 1,
-      [Define if you can turn off readline's signal handling.]),,
-    [
-#include <stdio.h> /* Must be first for Gnu Readline */
-#ifdef WITH_EDITLINE
-# include <editline/readline.h>
-#else
-# include <readline/readline.h>
-#endif
-    ])
+    # check for readline 2.2
+    AC_CHECK_DECL(rl_completion_append_character,
+      AC_DEFINE(HAVE_RL_COMPLETION_APPEND_CHARACTER, 1,
+        [Define if you have readline 2.2]),,
+      [
+      #include <stdio.h> /* Must be first for Gnu Readline */
+      #ifdef WITH_EDITLINE
+      # include <editline/readline.h>
+      #else
+      # include <readline/readline.h>
+      #endif
+      ])
 
-  AC_CHECK_LIB($LIBREADLINE, append_history,
-    AC_DEFINE(HAVE_RL_APPEND_HISTORY, 1,
-      [Define if readline supports append_history]),,$READLINE_LIBS)
-fi
+    AC_CHECK_DECL(rl_completion_suppress_append,
+      AC_DEFINE(HAVE_RL_COMPLETION_SUPPRESS_APPEND, 1,
+        [Define if you have rl_completion_suppress_append]),,
+      [
+      #include <stdio.h> /* Must be first for Gnu Readline */
+      #ifdef WITH_EDITLINE
+      # include <editline/readline.h>
+      #else
+      # include <readline/readline.h>
+      #endif
+      ])
+
+    # check for readline 4.0
+    AC_CHECK_LIB($LIBREADLINE, rl_pre_input_hook,
+      AC_DEFINE(HAVE_RL_PRE_INPUT_HOOK, 1,
+        [Define if you have readline 4.0]))
```

Comment:
Looks like this will create havoc for WASM/Emscripten static builds; multiple `-l` switches are passed to the compiler:

config.log:
```
29899 configure:21034: checking for rl_pre_input_hook in -lreadline
29900 configure:21059: gcc -o conftest     conftest.c -lreadline  -lreadline -lintl -ldl  >&5
29901 configure:21059: $? = 0
29902 configure:21069: result: yes
29903 configure:21080: checking for rl_completion_display_matches_hook in -lreadline
29904 configure:21105: gcc -o conftest     conftest.c -lreadline  -lreadline -lintl -ldl  >&5
29905 configure:21105: $? = 0
29906 configure:21115: result: yes
29907 configure:21126: checking for rl_resize_terminal in -lreadline
29908 configure:21151: gcc -o conftest     conftest.c -lreadline  -lreadline -lintl -ldl  >&5
```

Reply:
`WITH_SAVE_ENV` can't be nested :( But your `LIBS_SAVE` workaround should work well.

Reply:
Unfortunately, it does not work. They're still there.

Reply:
`AC_CHECK_LIB` might be prefixing the library before running the check.

Reply:
Yeah, we need a different approach here.

Reply:
Yeah. This reminds me I've got an open PR for the exact same issue for sqlite3 dependency detection :) I've tried various approaches, but I haven't found a neat way to solve it yet. (I haven't looked at it for a week either, though.)